src/iorecord.c

```diff
@@ -0,0 +1,168 @@
+                        exit(EXIT_FAILURE);
+                }
+        }
+
+        tmpfd = mkstemp(filepath);
+        if (!tmpfd) {
+                telem_perror("Error opening staging file");
+                close(tmpfd);
+                if (unlink(filepath)) {
+                        telem_perror("Error deleting staging file");
+                }
+                goto clean_exit;
+        }
+
+        // open file
+        tmpfile = fdopen(tmpfd, "a");
+        if (!tmpfile) {
+                telem_perror("Error opening temp stage file");
+                close(tmpfd);
+                if (unlink(filepath)) {
+                        telem_perror("Error deleting temp stage file");
+                }
+                goto clean_exit;
+        }
+
+        // write headers
+        for (int i = 0; i < NUM_HEADERS; i++) {
+                fprintf(tmpfile, "%s\n", headers[i]);
+        }
+
+        //write body
+        fprintf(tmpfile, "%s\n", body);
+        fflush(tmpfile);
+        fclose(tmpfile);
+
+clean_exit:
+
+        free(filepath);
+}
+
+bool read_record(char *fullpath, char *headers[], char **body)
+{
+        int i, ret = 0;
+        bool result = false;
+        FILE *fp = NULL;
+        long offset;
+        char line[LINE_MAX] = { 0 };
+        struct stat buf;
+        long size;
+
+        ret = stat(fullpath, &buf);
+        if (ret == -1) {
+                telem_log(LOG_ERR, "Unable to stat record %s in staging\n", fullpath);
+                return false;
+        }
+        size = buf.st_size;
+
+        fp = fopen(fullpath, "r");
+        if (fp == NULL) {
+                telem_log(LOG_ERR, "Unable to open file %s in staging\n", fullpath);
+                return false;
+        }
+
+        for (i = 0; i < NUM_HEADERS; i++) {
+                const char *header_name = get_header_name(i);
+                if (!_fgets(line, LINE_MAX, fp)) {
+                        telem_log(LOG_ERR, "Error while parsing staged record\n");
+                        fclose(fp);
+                        return false;
+                }
+                //Get rid of trailing newline
+                result = get_header(line, header_name, &headers[i]);
+                if (!result) {
+                        telem_log(LOG_ERR, "read_record: Incorrect"
+                                  " headers in record\n");
+                        goto read_error;
+                }
+        }
+
+        offset = ftell(fp);
+        if (offset == -1) {
+                telem_perror("Failed to get staged file offset of record payload");
+                result = false;
+                goto read_error;
+        }
+
+        size = size - offset + 1;
+        *body = malloc((size_t)size);
+        if (*body == NULL) {
+                telem_log(LOG_ERR, "Could not allocate memory for payload from staged file\n");
+                fclose(fp);
```

Comment:
Either set `fp` to `NULL` before the goto, or remove the `fclose()` here, because fp is already closed on line 164. Or else, the fp may be closed twice, and we want to avoid that.